File: ascii_fields/animations/storm.py
import math

from ..core import Animation, clamp, render_field
from ..noise import fbm


class JupiterStormAnimation(Animation):
  """Jupiter's Great Red Spot as a layered banded atmosphere with a rotating
  oval vortex, a dark eye, and turbulent streamers around it."""

  default_theme = "sunset"

  thresholds = [
    (0.08, " "), (0.18, "."), (0.30, ":"), (0.42, "-"), (0.54, "="),
    (0.66, "+"), (0.78, "*"), (0.90, "#"), (1.01, "@"),
  ]

  def render(self, width, height, elapsed, phase, options):
    ax = width / max(1, height * 2.0)
    cx = 0.08 * ax * math.sin(elapsed * 0.12)
    cy = 0.03 * math.sin(elapsed * 0.18)
    spin = elapsed * 0.55
    contrast = options.contrast
    # bands drift horizontally so the atmosphere is alive instead of static
    band_drift = elapsed * 0.06
    cos_spin = math.cos(spin)
    sin_spin = math.sin(spin)
    inv_rx = 1.0 / max(0.22, 0.36 * ax)
    grid = []

    for row in range(height):
      y = (row / max(1, height - 1) - 0.5) * 2.0
      # per-row band invariants (cheap row-only sines hoisted out of inner loop)
      band_base = 0.10 + 0.13 * math.sin(y * 18.0 + elapsed * 0.25)
      band_base += 0.07 * math.sin(y * 43.0 - elapsed * 0.15)
      shear = 0.12 * math.sin(y * 9.0 + elapsed * 0.2)
      wind_freq_x = 1.8 * (0.10 + 0.08 * math.sin(y * 7.0)) + 1.0  # constant per row
      dy = (y - cy) / 0.25
      line = []
      for col in range(width):
        x = (col / max(1, width - 1) - 0.5) * 2.0 * ax
        wind = fbm(x * 1.8 + band_drift, y * 5.0 + shear, octaves=3)
        value = band_base + 0.18 * wind

        dx = (x - cx) * inv_rx
        r = math.hypot(dx, dy)
        theta = math.atan2(dy, dx)
        oval = math.exp(-(r * r) * 1.35)
        wall = math.exp(-((r - 0.95) ** 2) / 0.028)
        eye = math.exp(-(r * r) / 0.12)
        swirl = 0.5 + 0.5 * math.sin(theta * 4.0 - spin + r * 8.5)
        turbulent = fbm(dx * 2.2 + cos_spin * 0.4, dy * 2.2 + sin_spin * 0.4, octaves=3)
        streamer = math.exp(-((abs(dy) - 0.75) ** 2) / 0.10)
        value += oval * (0.25 + 0.30 * swirl + 0.20 * turbulent)
        value += wall * (0.42 + 0.22 * swirl)
        value -= eye * 0.55              # dark eye of the storm
        value += streamer * max(0.0, 1.0 - abs(dx) * 0.9) * 0.18
        line.append(clamp(value * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/storm.rs
use rayon::prelude::*;

use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};
use crate::noise::fbm;

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "sunset" };
const TH: &[(f64, char)] = &[
  (0.08, ' '),
  (0.18, '.'),
  (0.30, ':'),
  (0.42, '-'),
  (0.54, '='),
  (0.66, '+'),
  (0.78, '*'),
  (0.90, '#'),
  (1.01, '@'),
];

pub struct Storm;

impl Animation for Storm {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width;
    let h = ctx.height;
    let ax = w as f64 / (h as f64 * 2.0).max(1.0);
    let cx = 0.08 * ax * (ctx.elapsed * 0.12).sin();
    let cy = 0.03 * (ctx.elapsed * 0.18).sin();
    let spin = ctx.elapsed * 0.55;
    let contrast = ctx.options.contrast;
    let band_drift = ctx.elapsed * 0.06;
    let cos_spin = spin.cos();
    let sin_spin = spin.sin();
    let inv_rx = 1.0 / (0.36 * ax).max(0.22);
    let mut grid = vec![0.0_f64; w * h];
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    let t = ctx.elapsed;
    grid.par_chunks_mut(w).enumerate().for_each(|(row, row_slice)| {
      let y = (row as f64 / dh - 0.5) * 2.0;
      let band_base = 0.10 + 0.13 * (y * 18.0 + t * 0.25).sin() + 0.07 * (y * 43.0 - t * 0.15).sin();
      let shear = 0.12 * (y * 9.0 + t * 0.2).sin();
      let dy = (y - cy) / 0.25;
      let abs_dy = dy.abs();
      let streamer_y = (-((abs_dy - 0.75).powi(2)) / 0.10).exp();
      for col in 0..w {
        let x = (col as f64 / dw - 0.5) * 2.0 * ax;
        let wind = fbm(x * 1.8 + band_drift, y * 5.0 + shear, 3);
        let mut value = band_base + 0.18 * wind;
        let dx = (x - cx) * inv_rx;
        let r = (dx * dx + dy * dy).sqrt();
        let theta = dy.atan2(dx);
        let oval = (-(r * r) * 1.35).exp();
        let wall = (-((r - 0.95).powi(2)) / 0.028).exp();
        let eye = (-(r * r) / 0.12).exp();
        let swirl = 0.5 + 0.5 * (theta * 4.0 - spin + r * 8.5).sin();
        let turbulent = fbm(dx * 2.2 + cos_spin * 0.4, dy * 2.2 + sin_spin * 0.4, 3);
        value += oval * (0.25 + 0.30 * swirl + 0.20 * turbulent);
        value += wall * (0.42 + 0.22 * swirl);
        value -= eye * 0.55;
        value += streamer_y * (1.0 - dx.abs() * 0.9).max(0.0) * 0.18;
        row_slice[col] = clamp(value * contrast);
      }
    });
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
